File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    N = len(jobs)
    
    idx, cur_time, pre_start = 0, 0, -1
    heap = []
    sum_ = 0
    while idx <= N-1:
        for job in jobs:
            if pre_start < job[0] <= cur_time:
                heapq.heappush(heap, [job[1], job[0]])
        if heap:
            job = heapq.heappop(heap)
            pre_start = cur_time
            cur_time += job[0]
            sum_ += cur_time - job[1]
            idx += 1
        else:
            cur_time += 1
    return sum_ // N

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<vector<int>> jobs) {
    int N = jobs.size();
    int Sum = 0;
    
    priority_queue<vector<int>> pq; //{소요시간, 요청시각}
    int j_cnt = 0; int cur_sec = 0; int pre_sec = -1;
    while (j_cnt < N)
    {
        for (auto &job : jobs)
        {
            if ((job[0] > pre_sec) && job[0] <= cur_sec)
            {
                pq.push({-job[1], -job[0]});
            }
        }
        if (!pq.empty())
        {
            vector<int> tjob = pq.top();
            pq.pop();
            pre_sec = cur_sec;
            cur_sec += (-tjob[0]);
            Sum += (cur_sec - (-tjob[1]));
            j_cnt++;
        }
        else
        {
            cur_sec++;
        }
    }
    
    return int(Sum / N);
}